mssw/Animal/lib/LivingCreature.pm

```diff
@@ -17,77 +19,171 @@ Version 0.01
 our $VERSION = '0.01';
 
 =head1 SYNOPSIS
-
-Quick summary of what the module does.
-
-Perhaps a little code snippet.
-
-    use LivingCreature;
-
-    my $foo = LivingCreature->new();
-    ...
-
+#
+#Quick summary of what the module does.
+#
+#Perhaps a little code snippet.
+#
+#    use LivingCreature;
+#
+#    my $foo = LivingCreature->new();
+#    ...
+#
 =head1 EXPORT
+#
+#A list of functions that can be exported.  You can delete this section
+#if you don't export anything, such as for a purely object-oriented module.
+#
+=head1 SUBROUTINES/METHODS
 
-A list of functions that can be exported.  You can delete this section
-if you don't export anything, such as for a purely object-oriented module.
+=head2 speak
+#
+#Allows a LivingCreature to speak
+#
+=cut
 
-=head1 SUBROUTINES/METHODS
+sub speak {
+    my $class = shift;
+    my ($say) = @_;
 
-=head2 function1
+    if ( defined $say ) {
+        print "a $class says: $say!\n";
+    }
+    else {
+        print "a $class goes ", $class->sound, "!\n";
+    }
+}
 
+=head2 sound
+#
+#Implement me!!!
+#
 =cut
 
-sub function1 {
+sub sound {
+    croak "Sound is unimplemented!!!";
 }
 
-=head2 function2
+=head2 default_colour
+#
+#returns the default colour of a LivingCreature. Override this where necessary.
+#
+=cut
 
+sub default_colour {
+    return 'dirty';
+}
+
+=head2 new
+#
+#Creates an instance
+#
 =cut
 
-sub function2 {
+sub new {
+    my $class  = shift;
+    my ($name) = @_;
+    my $self   = { name => $name, };
+
+    return bless $self, $class;
 }
 
-=head1 AUTHOR
+=head2 name
+#
+#can be used to get/set the name of a LivingCreature
+#
+=cut
 
-Michael Wambeek, C<< <mikewambeek at hotmail.co.uk> >>
+sub name {
+    my $self = shift;
 
-=head1 BUGS
+    # Was this method called statically?
+    if ( !ref $self ) {
 
-Please report any bugs or feature requests to C<bug-. at rt.cpan.org>, or through
-the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=.>.  I will be notified, and then you'll
-automatically be notified of progress on your bug as I make changes.
+        # It was! well we cant set anything so we'll just die
+        if (@_) {
+            croak "You can't set a name on a default $self";
+        }
 
+        return $self;
```

Comment:
It would chain but in the context that the caller is expecting a getter. This is fine if you want the name of the animal class to just be the animal name, though I think the book went with something like `return "an unnamed $self";`.